// tap-bark/src/stream/stream.js
"use strict";
var readline = require("readline");
var through = require("through2");
var Stream = (function () {
    function Stream() {
        this._stream = through();
    }
    Object.defineProperty(Stream.prototype, "stream", {
        get: function () {
            return this._stream;
        },
        set: function (value) {
            this._stream = value;
        },
        enumerable: true,
        configurable: true
    });
    Stream.prototype.writeLine = function (message) {
        this._stream.write(message + "\n");
    };
    Stream.prototype.write = function (message) {
        this._stream.write(message);
    };
    Stream.prototype.moveCursor = function (x, y) {
        readline.moveCursor(this._stream, x, y);
    };
    Stream.prototype.cursorTo = function (x, y) {
        readline.cursorTo(this._stream, x, y);
    };
    Stream.prototype.clearLine = function () {
        readline.clearLine(this._stream, 0);
    };
    Stream.prototype.getUnderlyingStream = function () {
        return this._stream;
    };
    return Stream;
}());
exports.Stream = Stream;
